### PR TITLE
Remove SnapshotPageProxy hard coded dependancy

### PR DESCRIPTION
**Pull request by FabienD**
According what I've explained in the issue #551, I've built this PR in order to make SnapshotPageProxy customisable in an easy way (I think).

### CODE REVIEW

File: Entity/Transformer.php [hunk 1]
```diff
@@ -33,6 +32,8 @@ class Transformer implements TransformerInterface
 
     protected $blockManager;
 
+    protected $snapshotPageProxyClass;
```
Comment (core23) [inline Entity/Transformer.php]: phpdoc missing

File: Entity/SnapshotManager.php [hunk 2]
```diff
@@ -36,6 +35,8 @@ class SnapshotManager extends BaseEntityManager implements SnapshotManagerInterf
      */
     protected $templates = array();
 
+    protected $snapshotPageProxyClass;
```
Comment (core23) [inline Entity/SnapshotManager.php]: phpdoc missing

File: Entity/SnapshotManager.php [hunk 3]
```diff
@@ -43,11 +44,12 @@ class SnapshotManager extends BaseEntityManager implements SnapshotManagerInterf
      * @param ManagerRegistry $registry  An entity manager instance
      * @param array           $templates An array of templates
      */
-    public function __construct($class, ManagerRegistry $registry, $templates = array())
+    public function __construct($class, $snapshotPageProxyClass, ManagerRegistry $registry, $templates = array())
```
Comment (OskarStark) [inline Entity/SnapshotManager.php]: This is BC

File: CmsManager/CmsSnapshotManager.php [hunk 4]
```diff
@@ -36,13 +35,20 @@ class CmsSnapshotManager extends BaseCmsPageManager
...
      * @param SnapshotManagerInterface $snapshotManager
      * @param TransformerInterface     $transformer
+     * @param string                   $snapshotPageProxyClass Namespace of SnaphostPageProxy class
      */
-    public function __construct(SnapshotManagerInterface $snapshotManager, TransformerInterface $transformer)
+    public function __construct(SnapshotManagerInterface $snapshotManager, TransformerInterface $transformer, $snapshotPageProxyClass = 'Sonata\PageBundle\Model\SnapshotPageProxy')
     {
         $this->snapshotManager = $snapshotManager;
         $this->transformer = $transformer;
+        $this->snapshotPageProxyClass = $snapshotPageProxyClass;
```
Comment (Soullivaneuh) [inline CmsManager/CmsSnapshotManager.php]: Is this should be mandatory?

In this case, you should throw a deprecated error.

File: Model/SnapshotPageProxyInterface.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+namespace Sonata\PageBundle\Model;
+
+/**
+ * SnapshotPageProxyInterface.
+ *
+ * @author Fabien D. <fabien@myprod.net>
+ */
+interface SnapshotPageProxyInterface extends PageInterface, \Serializable
+{
+}
```
Comment (Soullivaneuh) [inline Model/SnapshotPageProxyInterface.php]: Maybe some methods could be added here?

**Comment by Soullivaneuh** [comment]
> I've built this PR in order to make SnapshotPageProxy customisable in an easy way (I think).

Could you please elaborate? Do you have some concrete cases where this class should be extended / overridden?

**Comment by FabienD** [comment]
Hi,

I've made this PR, just because, it's the only way I found to add a new property on the Page entity (cf. #551). I want to have the possibility to set a specific css class by page.

Thx.

File: CmsManager/CmsSnapshotManager.php (hunk 4 above)
Comment (FabienD) [inline CmsManager/CmsSnapshotManager.php]: Yes, it is. A default value is provided in the constructor.

File: Model/SnapshotPageProxyInterface.php (hunk 5 above)
Comment (FabienD) [inline Model/SnapshotPageProxyInterface.php]: I've done that in order to test the type of the object in "Twig/Extension/PageExtension.php"

**Comment by Soullivaneuh** [comment]
@rande what do you think?

**Comment by FabienD** [comment]
No feedback ?
Thx.

**Comment by rande** [comment]
I will prefer a factory to avoid injecting class name.

**Comment by FabienD** [comment]
Ok, I'll try to do that. Thank you.

**Comment by FabienD** [comment]
Hi,

I've refactored the PR in order to use a factory to instantiate the **SnapshotProxy** class. Now, the **SnapshotManager** takes a 4th argument, the **SnapshotProxyFactory**. I'm not confortable with "tests", I've updated the **SnapshotManagerTest** but I don't know how to fix the failed test in **CmsSnapshotManagerTest**. I need help to fix the failed tests.

In addition, I've removed the **getPageByName** method from the **SnapshotManager**, this method seems not to be used, the **snapshotProxy** is instantiated with only 2 args, the class needs 3.

What do you think about this refactoring ?

Thx.

File: Entity/SnapshotManager.php [hunk 6]
```diff
@@ -37,17 +39,24 @@ class SnapshotManager extends BaseEntityManager implements SnapshotManagerInterf
...
-     * @param string          $class     Namespace of entity class
-     * @param ManagerRegistry $registry  An entity manager instance
-     * @param array           $templates An array of templates
+     * @param string                   $class                    Namespace of entity class
+     * @param ManagerRegistry          $registry                 An entity manager instance
+     * @param array                    $templates                An array of templates
+     * @param SnapshotPageProxyFactory $snapshotPageProxyFactory SnapshotPageProxy factory
      */
-    public function __construct($class, ManagerRegistry $registry, $templates = array())
+    public function __construct($class, ManagerRegistry $registry, $templates = array(), SnapshotPageProxyFactory $snapshotPageProxyFactory)
```
Comment (core23) [inline Entity/SnapshotManager.php]: This is a BC break

File: Entity/SnapshotManager.php [hunk 7]
```diff
@@ -140,34 +149,6 @@ public function findEnableSnapshot(array $criteria)
     }
 
     /**
-     * return a page with the given routeName.
-     *
-     * @param string $routeName
-     *
-     * @return PageInterface|false
-     */
-    public function getPageByName($routeName)
```
Comment (core23) [inline Entity/SnapshotManager.php]: BC break

File: Entity/Transformer.php [hunk 8]
```diff
@@ -52,13 +51,14 @@ class Transformer implements TransformerInterface
      * @param PageManagerInterface     $pageManager
      * @param BlockManagerInterface    $blockManager
      * @param RegistryInterface        $registry
+     * @param string                   $snapshotPageProxyClass Namespace of SnapshotPageProxy class
      */
```
Comment (core23) [inline Entity/Transformer.php]: Parameter is missing in the method header ;)

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.

**Pull request closed without merging**